src/net/p2p.rs

```diff
@@ -1625,46 +1633,54 @@ impl PeerNetwork {
-                }
-            },
-            PeerNetworkWorkState::BlockInvSync => {
-                // synchronize peer block inventories 
-                if self.do_network_inv_sync(burndb)? {
-                    // proceed to get blocks
-                    self.work_state = PeerNetworkWorkState::BlockDownload;
-                }
-            },
-            PeerNetworkWorkState::BlockDownload => {
-                // go fetch blocks
-                match dns_client_opt {
-                    Some(ref mut dns_client) => {
-                        if self.do_network_block_download(burndb, chainstate, *dns_client, network_result)? {
-                            // advance work state
+        while !do_prune {
+            let cur_state = self.work_state;
+            match self.work_state {
+                PeerNetworkWorkState::NeighborWalk => {
+                    // walk the peer graph and deal with new/dropped connections
+                    if self.do_network_neighbor_walk()? {
+                        // proceed to synchronize block invs
+                        self.work_state = PeerNetworkWorkState::BlockInvSync;
+                    }
+                },
+                PeerNetworkWorkState::BlockInvSync => {
+                    // synchronize peer block inventories 
+                    if self.do_network_inv_sync(burndb)? {
+                        // proceed to get blocks
+                        self.work_state = PeerNetworkWorkState::BlockDownload;
+                    }
+                },
+                PeerNetworkWorkState::BlockDownload => {
+                    // go fetch blocks
+                    match dns_client_opt {
+                        Some(ref mut dns_client) => {
+                            if self.do_network_block_download(burndb, chainstate, *dns_client, network_result)? {
+                                // advance work state
+                                self.work_state = PeerNetworkWorkState::Prune;
+                            }
+                        },
+                        None => {
+                            // skip this step -- no DNS client available
+                            test_debug!("{:?}: no DNS client provided; skipping block download", &self.local_peer);
                             self.work_state = PeerNetworkWorkState::Prune;
                         }
-                    },
-                    None => {
-                        // skip this step -- no DNS client available
-                        test_debug!("{:?}: no DNS client provided; skipping block download", &self.local_peer);
-                        self.work_state = PeerNetworkWorkState::Prune;
                     }
+                },
+                PeerNetworkWorkState::Prune => {
+                    // clear out neighbor connections after we finish sending
+                    if self.do_prune {
+                        do_prune = true;
+                        self.do_prune = false;
```

Comment:
Will this function loop until self.do_prune is set to true by some call to `do_network_neighbor_walk` gets a `walk_result` that requests a prune?

Reply:
ah, typo.  Meant to just make the loop break once the Prune state is reached

Reply:
Lemme change that and re-test it.  I just wanted the loop to break once the network state did one round of work.

Reply:
Addressed in e0349d2